subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2127 @@
+		return 1;
+	}
+
+	strm->vbuf_offset = next_vbuf_offset;
+	vbuf->line_offset = next_line_offset;
+	return 0;
+}
+
+static void uvc_worker(struct k_work *work)
+{
+	struct uvc_stream *strm = CONTAINER_OF(work, struct uvc_stream, work);
+	struct video_buffer *vbuf;
+	int ret;
+
+	if (!atomic_test_bit(&strm->state, CLASS_ENABLED) ||
+	    !atomic_test_bit(&strm->state, CLASS_READY)) {
+		LOG_DBG("not ready");
+		return;
+	}
+
+	while ((vbuf = k_fifo_peek_head(&strm->fifo_in)) != NULL) {
+		ret = uvc_queue_vbuf(strm, vbuf);
+		if (ret < 0) {
+			LOG_DBG("Could not transfer %p for now", vbuf);
+			break;
+		}
+		if (ret == 1) {
+			LOG_DBG("Video buffer %p transferred, removing from the queue", vbuf);
+			k_fifo_get(&strm->fifo_in, K_NO_WAIT);
+		}
+	}
+}
+
+static void uvc_enable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct uvc_data *data = dev->data;
+
+	for (struct uvc_stream *strm = data->streams; strm->dev != NULL; strm++) {
+		/* Catch-up with buffers that might have been delayed */
+		atomic_set_bit(&strm->state, CLASS_ENABLED);
+		k_work_submit(&strm->work);
+	}
+}
+
+static void uvc_disable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct uvc_data *data = dev->data;
+
+	for (struct uvc_stream *strm = data->streams; strm->dev != NULL; strm++) {
+		atomic_clear_bit(&strm->state, CLASS_ENABLED);
+	}
+}
+
+static const struct usbd_class_api uvc_class_api = {
+	.enable = uvc_enable,
+	.disable = uvc_disable,
+	.request = uvc_request,
+	.update = uvc_update,
+	.control_to_host = uvc_control_to_host,
+	.control_to_dev = uvc_control_to_dev,
+	.init = uvc_init,
+	.get_desc = uvc_get_desc,
+};
+
+static int uvc_enqueue(const struct device *dev, enum video_endpoint_id ep,
+		       struct video_buffer *vbuf)
+{
+	struct uvc_stream *strm;
+	int ret;
+
+	ret = uvc_get_stream(dev, ep, &strm);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_fifo_put(&strm->fifo_in, vbuf);
+	k_work_submit(&strm->work);
+	return 0;
+}
+
+static int uvc_dequeue(const struct device *dev, enum video_endpoint_id ep,
+		       struct video_buffer **vbuf, k_timeout_t timeout)
+{
+	struct uvc_stream *strm;
+	int ret;
+
+	ret = uvc_get_stream(dev, ep, &strm);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*vbuf = k_fifo_get(&strm->fifo_out, timeout);
+	if (*vbuf == NULL) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+static int uvc_get_format(const struct device *dev, enum video_endpoint_id ep,
+			  struct video_format *fmt)
+{
+	struct uvc_stream *strm;
+	int ret;
+
+	ret = uvc_get_stream(dev, ep, &strm);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!atomic_test_bit(&strm->state, CLASS_ENABLED) ||
+	    !atomic_test_bit(&strm->state, CLASS_READY)) {
+		return -EAGAIN;
+	}
+
+	return video_get_format(strm->video_dev, VIDEO_EP_OUT, fmt);
+}
+
+static int uvc_stream_start(const struct device *dev)
+{
+	/* TODO: resume the stream after it was interrupted if needed */
```

Comment:
The stream start as a reaction to the host starting it: UVC starts the application stream rather than the other way around.

I am actually wondering what would be the best approach for a video component that is *driving* a video pipeline.

Reply:
The `uvc_stream_start/stop()` left as TODO were expected to block the stream and let it pursue after the host started the pipeline. An extra "pause" that the application can apply to the stream.